packages/flutter/lib/src/material/image_icon.dart

```diff
@@ -0,0 +1,98 @@
+// Copyright 2015 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:flutter/services.dart';
+import 'package:flutter/widgets.dart';
+
+import 'icons.dart';
+import 'icon_button.dart';
+import 'icon_theme.dart';
+import 'icon_theme_data.dart';
+
+/// An icon that comes from an [ImageProvider], e.g. an [AssetImage].
+///
+/// See also:
+///
+///  * [IconButton], for interactive icons.
+///  * [IconTheme], which provides ambient configuration for icons.
+///  * [Icon] and [Icons], for icons from the material design library.
+class ImageIcon extends StatelessWidget {
+  /// Creates an image icon.
+  ///
+  /// The [size] and [color] default to the value given by the current [IconTheme].
+  const ImageIcon(this.image, {
+    Key key,
+    this.size,
+    this.color
+  }) : super(key: key);
+
+  /// The image to display as the icon.
+  ///
+  /// The icon can be null, in which case the widget will render as an empty
+  /// space of the specified [size].
+  final ImageProvider image;
+
+  /// The size of the icon in logical pixels.
+  ///
+  /// Icons occupy a square with width and height equal to size.
+  ///
+  /// Defaults to the current [IconTheme] size, if any. If there is no
+  /// [IconTheme], or it does not specify an explicit size, then it defaults to
+  /// 24.0.
+  final double size;
+
+  /// The color to use when drawing the icon.
+  ///
+  /// Defaults to the current [IconTheme] color, if any. If there is
+  /// no [IconTheme], then it defaults to not recolorizing the image.
+  ///
+  /// The image will additionally be adjusted by the opacity of the current
+  /// [IconTheme], if any.
+  final Color color;
+
+  @override
+  Widget build(BuildContext context) {
+    final IconThemeData iconTheme = const IconThemeData.fallback().merge(IconTheme.of(context));
+
+    final double iconSize = size ?? iconTheme.size;
+
+    if (image == null)
+      return new SizedBox(width: iconSize, height: iconSize);
+
+    final double iconOpacity = iconTheme.opacity;
+    final Color iconColor = color ?? iconTheme.color;
+
+    Widget result = new Image(
+      image: image,
+      width: iconSize,
+      height: iconSize,
+      color: iconColor,
+      fit: ImageFit.scaleDown,
+      alignment: FractionalOffset.center
+    );
+
+    if (iconOpacity != 1.0) {
+      result = new Opacity(
```

Comment:
It would be much more efficient to do this with a color filter on the image.  We already have a `color` parameter on `Image` that turns into a `ColorFilter`.  We should probably upgrade that to a `ColorFilter` and use that to create the opacity.

We did it this goofy way before because `ColorFilter` was a host object (and therefore expensive to create).  I guess we still need to add `operator==` to https://github.com/flutter/engine/blob/master/flutter/lib/ui/painting.dart#L836 ...

If that's too much of a yak shave, please feel free to land it like this with a TODO comments and you or I can shave the yak in another patch.